Refactor script text input

AllyTally opened the pull request:
## Changes

This PR adds an input system with selection, copying, pasting, and all of that fun stuff, to the script editor.

NOTE: I'm still new to text input, so if someone else sees issues then feel free to contribute.

## Legal Stuff:

By submitting this pull request, I confirm that...

- [x] My changes may be used in a future commercial release of VVVVVV
- [x] I will be credited in a `CONTRIBUTORS` file and the "GitHub Friends"
  section of the credits for all of said releases, but will NOT be compensated
  for these changes

AllyTally committed:
Add base text input system

AllyTally committed:
Make script editor use new input system

The old input system still exists for one-line inputs, but the script
editor really needs the upgrade. This commit fixes a few bugs with the
input system, and makes the script editor use it.

AllyTally committed:
Add more text input keybinds

This commit adds HOME, END, PAGEUP, PAGEDOWN, CTRL+LEFT/RIGHT, DEL,
CTRL+BACKSPACE and CTRL+DEL. It also rewrites the text insertion
function to not use stringstreams.

AllyTally committed:
Move everything else to the new textinput system

This commit makes the system support single-line text input, and moves
every remaining text input to it. This disables selection and cursor
movement for single line input, since drawing that seems a bit hard
right now.

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+        if (cursor_pos.y > 0) {
+            cursor_pos.y--;
+            cursor_pos.x = cursor_x_tallest;
+            if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp: In function ‘void TextInput::move_cursor_up()’:
VVVVVV/desktop_version/src/TextInput.cpp:305:30: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  305 |             if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
      |                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+        if (cursor_pos.y < get_lines() - 1) {
+            cursor_pos.y++;
+            cursor_pos.x = cursor_x_tallest;
+            if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp: In function ‘void TextInput::move_cursor_down()’:
VVVVVV/desktop_version/src/TextInput.cpp:321:30: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  321 |             if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
      |                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+
+    bool is_word_part(char character)
+    {
+        return character > 0x7F || isalpha(character) || isdigit(character) || character == '_' || character == '-';
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp: In function ‘bool TextInput::is_word_part(char)’:
VVVVVV/desktop_version/src/TextInput.cpp:333:26: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  333 |         return character > 0x7F || isalpha(character) || isdigit(character) || character == '_' || character == '-';
      |                ~~~~~~~~~~^~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+            cursor_pos.x = cursor_select_pos.x;
+        }
+
+        if (cursor_pos.x >= UTF8_total_codepoints(get_line(cursor_pos.y)) && cursor_pos.y == get_lines() - 1) {
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp: In function ‘void TextInput::move_cursor_right()’:
VVVVVV/desktop_version/src/TextInput.cpp:395:26: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  395 |         if (cursor_pos.x >= UTF8_total_codepoints(get_line(cursor_pos.y)) && cursor_pos.y == get_lines() - 1) {
      |             ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+                }
+                SDL_free(character);
+                cursor_pos.x++;
+                if (cursor_pos.x >= UTF8_total_codepoints(get_line(cursor_pos.y)) && cursor_pos.y < get_lines() - 1)
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp:417:34: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  417 |                 if (cursor_pos.x >= UTF8_total_codepoints(get_line(cursor_pos.y)) && cursor_pos.y < get_lines() - 1)
      |                     ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+        else
+        {
+
+            if (cursor_pos.x < UTF8_total_codepoints(get_line(cursor_pos.y))) {
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp:428:30: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  428 |             if (cursor_pos.x < UTF8_total_codepoints(get_line(cursor_pos.y))) {
      |                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+            return;
+        }
+
+        if (cursor_pos.x == UTF8_total_codepoints(get_line(cursor_pos.y)))
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp: In function ‘void TextInput::delete_key()’:
VVVVVV/desktop_version/src/TextInput.cpp:545:26: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  545 |         if (cursor_pos.x == UTF8_total_codepoints(get_line(cursor_pos.y)))
      |             ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+                if (cursor_pos.y < 0) {
+                    cursor_pos.y = 0;
+                }
+                if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp: In function ‘void TextInput::handle_events(SDL_Event)’:
VVVVVV/desktop_version/src/TextInput.cpp:701:34: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  701 |                 if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
      |                     ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,729 @@
+                if (cursor_pos.y >= get_lines()) {
+                    cursor_pos.y = get_lines() - 1;
+                }
+                if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
```
```c++
VVVVVV/desktop_version/src/TextInput.cpp:711:34: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  711 |                 if (cursor_pos.x > UTF8_total_codepoints(get_line(cursor_pos.y))) {
      |                     ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

InfoTeddy commented on desktop_version/src/Editor.cpp:
```diff
@@ -1583,10 +1579,89 @@ void editorrender(void)
+                const SDL_Color color = graphics.getRGB(123, 111, 218);
+                const int x = TextInput::cursor_select_pos.x;
+                const int y = TextInput::cursor_select_pos.y;
+                const int w = TextInput::cursor_pos.x - x;
```
```c++
/A/p/r/VVVVVV/desktop_version/src/Editor.cpp: In function ‘void editorrender()’:
VVVVVV/desktop_version/src/Editor.cpp:1590:27: warning: unused variable ‘w’ [-Wunused-variable]
 1590 |                 const int w = TextInput::cursor_pos.x - x;
      |                           ^
```

InfoTeddy reviewed (changes requested):
Here are the warnings I get.

Also, please fix your brace style. There are several places where you use same-line brace instead of next-line brace.

AllyTally committed:
Fix style and warnings

Co-authored-by: Misa Elizabeth Kai <infoteddy@infoteddy.info>

InfoTeddy commented on desktop_version/src/Editor.cpp:
```diff
@@ -1583,10 +1579,89 @@ void editorrender(void)
+            if (TextInput::selecting)
+            {
+                const SDL_Color color = graphics.getRGB(123, 111, 218);
+                const int x = TextInput::cursor_select_pos.x;
```
```c++
VVVVVV/desktop_version/src/Editor.cpp: In function ‘void editorrender()’:
VVVVVV/desktop_version/src/Editor.cpp:1589:27: warning: unused variable ‘x’ [-Wunused-variable]
 1589 |                 const int x = TextInput::cursor_select_pos.x;
      |                           ^
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,766 @@
+#include "UtilityClass.h"
+#include "Vlogging.h"
+
+namespace TextInput {
```
```suggestion
namespace TextInput
{
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,766 @@
+    void move_cursor_up(void)
+    {
+        bool reset = process_selection();  // Only returns true if you don't hold shift
+        if (reset && (cursor_pos.y > cursor_select_pos.y))
```
```suggestion
        if (reset && cursor_pos.y > cursor_select_pos.y)
```

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,766 @@
+    void move_cursor_down(void)
+    {
+        bool reset = process_selection();  // Only returns true if you don't hold shift
+        if (reset && (cursor_pos.y < cursor_select_pos.y))
```
```suggestion
        if (reset && cursor_pos.y < cursor_select_pos.y)
```

InfoTeddy reviewed (changes requested):
A warning and some style stuff.

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,767 @@
+
+    bool is_word_part(char character)
+    {
+        return (short) character < 0 || (short) character > 0x7F || isalpha(character) || isdigit(character) || character == '_' || character == '-';
```
Also, I just realized - you need to use `SDL_isdigit` and `SDL_isalpha` here.

InfoTeddy commented on desktop_version/src/TextInput.cpp:
```diff
@@ -0,0 +1,767 @@
+#include <SDL.h>
+#include <algorithm>
+#include <string>
+#include <sstream>
+#include <vector>
+#include <iostream>
+
+#include "KeyPoll.h"
+#include "TextInput.h"
+#include "UTF8.h"
+#include "UtilityClass.h"
+#include "Vlogging.h"
```
1. Why `#include` `<algorithm>`, `<sstream>`, and `<iostream>`? Do you ever make use of any of them? If not, then remove them.
2. `#include "TextInput.h"` should be at the very top, separated by a blank line, since this file (`TextInput.cpp`) is the file that implements it. This is how other files in the codebase do it.

AllyTally committed:
Abstract script text printing to TextInput func

Drawing selection & the cursor was very hardcoded before, so why not
let other things use it as well?

This also fixes a bug where selected text is drawn above the text's Y
position.

AllyTally committed:
Centered text support by Misa

This commit also fixes broken selection copying.

NyakoFox commented:
note to self: apparently this breaks terminals/script boxes making new scripts, though a fix is simple (stop using key.keybuffer in input_submitted())

InfoTeddy commented:
Conflicts need to be resolved.

NyakoFox commented:
Closing this because it's very stale and could use a lot more work.
